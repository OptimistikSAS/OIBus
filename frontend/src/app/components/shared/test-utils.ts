import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };
