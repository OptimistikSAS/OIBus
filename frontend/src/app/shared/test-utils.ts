import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';
import { Page } from './types';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };

export function toPage<T>(content: Array<T>, totalElements: number = content.length, num = 0, size = Math.max(20, totalElements)): Page<T> {
  return {
    content,
    totalElements,
    number: num,
    size,
    totalPages: Math.ceil(totalElements / size)
  };
}

export function emptyPage<T>(size = 20): Page<T> {
  return {
    totalElements: 0,
    content: [] as Array<T>,
    size,
    totalPages: 0,
    number: 0
  };
}
