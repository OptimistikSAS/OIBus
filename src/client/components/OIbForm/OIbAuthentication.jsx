import React from 'react'
import PropTypes from 'prop-types'
import { Row, Col } from 'reactstrap'

import OIbText from './OIbText.jsx'
import OIbPassword from './OIbPassword.jsx'
import OIbSelect from './OIbSelect.jsx'
import OIbTitle from './OIbTitle.jsx'
import { notEmpty, hasLengthBetween } from '../../../services/validation.service'

/*
  OIBAuthentication is a form reused in several places. Can manage user/password (default)
  or accessKey/secretKey as well as authentication type.
*/
const OIbAuthentication = ({ value, name, onChange, mode, label }) => {
  const handleChange = (attributeName, newValue, valid) => {
    onChange(`${name}.${attributeName}`, newValue, valid)
  }
  const validation = {
    accessKey: notEmpty(),
    secretKey: hasLengthBetween(0, 256),
    username: notEmpty(),
    password: hasLengthBetween(0, 256),
  }
  return [
    <OIbTitle label={label} key="title">
      <div>
        <p>Authentication parameters</p>
        <p>Please fill the user and password to connect this application</p>
      </div>
    </OIbTitle>,
    mode === 'user' ? (
      [
        <Row key="type">
          <Col md="2">
            <OIbSelect
              label="Type"
              onChange={handleChange}
              value={value.type}
              options={['Basic']}
              defaultValue="Basic"
              name="type"
            />
          </Col>
        </Row>,
        <Row key="user">
          <Col md="4">
            <OIbText
              label="User name"
              onChange={handleChange}
              value={value.username}
              valid={validation.username}
              name="username"
            />
          </Col>
          <Col md="4">
            <OIbPassword
              label="Password"
              onChange={handleChange}
              value={value.password}
              valid={validation.password}
              name="password"
            />
          </Col>
        </Row>,
      ]
    ) : (
      <Row key="accessKey">
        <Col md="4">
          <OIbText
            label="Access Key"
            onChange={handleChange}
            value={value.accessKey}
            valid={validation.accessKey}
            name="accessKey"
          />
        </Col>
        <Col md="4">
          <OIbPassword
            label="Secret Key"
            onChange={handleChange}
            value={value.secretKey}
            valid={validation.secretKey}
            name="secretKey"
          />
        </Col>
      </Row>
    ),
  ]
}

OIbAuthentication.propTypes = {
  onChange: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.object,
  mode: PropTypes.oneOf(['accessKey', 'user']),
  label: PropTypes.string,
}

OIbAuthentication.defaultProps = {
  value: { type: 'basic', username: '', password: '', accessKey: '', secretKey: '' },
  mode: 'user',
  label: 'Authentication',
}

export default OIbAuthentication
