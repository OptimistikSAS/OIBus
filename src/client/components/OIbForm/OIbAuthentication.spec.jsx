import React from 'react'
import ReactDOM from 'react-dom'
import { act, Simulate } from 'react-dom/test-utils'


import OIbAuthentication from './OIbAuthentication.jsx'

let container
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

describe('OIbAuthentication', () => {
  test('check Authentication with default mode (user/password)', () => {
    act(() => {
      ReactDOM.render(<OIbAuthentication
        value={{ type: 'basic', username: 'user', password: 'pass' }}
        name="name"
        onChange={() => (1)}
      />, container)
    })
    // need to replace random id and toggler with fixed values
    // so the snapshot can be compared between test runs
    const button = container.querySelector('.close')
    const div = container.querySelector('.collapse')
    button.id = 'id'
    div.setAttribute('toggler', 'toggler')
    expect(container).toMatchSnapshot()
  })
  test('check Authentication with  mode accessKey', () => {
    act(() => {
      ReactDOM.render(<OIbAuthentication
        value={{ type: 'basic', accessKey: 'aaaaaa', secretKey: 'sssss' }}
        name="name"
        onChange={() => (1)}
        mode="accessKey"
      />, container)
    })
    // need to replace random id and toggler with fixed values
    // so the snapshot can be compared between test runs
    const button = container.querySelector('.close')
    const div = container.querySelector('.collapse')
    button.id = 'id'
    div.setAttribute('toggler', 'toggler')
    expect(container).toMatchSnapshot()
  })
  test('check the returned value', () => {
    const onChangeMock = jest.fn()
    act(() => {
      ReactDOM.render(<OIbAuthentication
        value={{ type: 'basic', username: 'user', password: 'pass' }}
        name="anyname"
        onChange={onChangeMock}
      />, container)
    })
    const userField = container.querySelector('#username')
    const passwordField = container.querySelector('#password')
    Simulate.change(userField, { target: { value: 'toto' } })
    expect(onChangeMock).toBeCalledWith('anyname.username', 'toto', null)
    Simulate.change(passwordField, { target: { value: 'passpass' } })
    expect(onChangeMock).toBeCalledWith('anyname.password', '{{notEncrypted}}passpass', null)
    // Test second render and componentDidUpdate
  })
})
