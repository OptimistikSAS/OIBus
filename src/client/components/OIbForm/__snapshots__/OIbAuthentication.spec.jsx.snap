// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OIbAuthentication check Authentication with  mode accessKey 1`] = `
<div>
  <div
    class="row"
  >
    <h5>
      Authentication
      <button
        aria-label="Close"
        class="close"
        id="id"
      >
        <svg
          class="oi-help"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
          />
        </svg>
      </button>
    </h5>
  </div>
  <div
    class="container-fluid"
  >
    <div
      class="collapse"
      toggler="toggler"
    >
      <div
        class="row"
        style="margin-bottom: 15px;"
      >
        <div
          class="col"
        >
          <div>
            <p>
              Authentication parameters
            </p>
            <p>
              Please fill the user and password to connect this application
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="col-md-4"
    >
      <div
        class="form-group"
      >
        <label
          class=""
          for="accessKey"
        >
          Access Key
        </label>
        <input
          class="oi-form-input form-control"
          id="accessKey"
          name="accessKey"
          type="text"
          value="aaaaaa"
        />
        <div
          class="invalid-feedback"
        />
      </div>
    </div>
    <div
      class="col-md-4"
    >
      <div
        class="form-group"
      >
        <label
          class=""
          for="secretKey"
        >
          Secret Key
        </label>
        <div
          class="input-group"
        >
          <input
            class="oi-form-input form-control"
            id="secretKey"
            name="secretKey"
            type="password"
            value="sssss"
          />
        </div>
        <div
          class="invalid-feedback"
          style="display: inline;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`OIbAuthentication check Authentication with default mode (user/password) 1`] = `
<div>
  <div
    class="row"
  >
    <h5>
      Authentication
      <button
        aria-label="Close"
        class="close"
        id="id"
      >
        <svg
          class="oi-help"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
          />
        </svg>
      </button>
    </h5>
  </div>
  <div
    class="container-fluid"
  >
    <div
      class="collapse"
      toggler="toggler"
    >
      <div
        class="row"
        style="margin-bottom: 15px;"
      >
        <div
          class="col"
        >
          <div>
            <p>
              Authentication parameters
            </p>
            <p>
              Please fill the user and password to connect this application
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="col-md-2"
    >
      <div
        class="form-group"
      >
        <label
          class=""
          for="type"
        >
          Type
        </label>
        <select
          class="oi-form-input form-control"
          id="type"
          name="type"
        >
          <option
            value="Basic"
          >
            Basic
          </option>
        </select>
        <div
          class="invalid-feedback"
        />
      </div>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="col-md-4"
    >
      <div
        class="form-group"
      >
        <label
          class=""
          for="username"
        >
          User name
        </label>
        <input
          class="oi-form-input form-control"
          id="username"
          name="username"
          type="text"
          value="user"
        />
        <div
          class="invalid-feedback"
        />
      </div>
    </div>
    <div
      class="col-md-4"
    >
      <div
        class="form-group"
      >
        <label
          class=""
          for="password"
        >
          Password
        </label>
        <div
          class="input-group"
        >
          <input
            class="oi-form-input form-control"
            id="password"
            name="password"
            type="password"
            value="pass"
          />
        </div>
        <div
          class="invalid-feedback"
          style="display: inline;"
        />
      </div>
    </div>
  </div>
</div>
`;
